activiti-core/activiti-engine/src/test/java/org/activiti/engine/test/history/HistoricTaskInstanceTest.java

```diff
@@ -605,6 +604,62 @@ public void testHistoricIdentityLinksOnTask() throws Exception {
             .satisfies(ae -> assertThat(ae.getObjectClass()).isEqualTo(HistoricTaskInstance.class));
     }
 
+    @Deployment(resources = {"org/activiti/engine/test/history/HistoricTaskInstanceTest.testHistoricIdentityLinksOnTask.bpmn20.xml"})
+    public void testHistoricIdentityLinksOnTaskWithDetails() throws Exception {
+        ProcessInstance processInstance = runtimeService.startProcessInstanceByKey("historicIdentityLinks");
+        Task task = taskService.createTaskQuery().processInstanceId(processInstance.getId()).singleResult();
+        assertThat(task).isNotNull();
+
+        // Set additional identity-link not coming from process
+        taskService.addUserIdentityLink(task.getId(),
+            "gonzo",
+            "actor",
+            "details".getBytes());
+        assertThat(taskService.getIdentityLinksForTask(task.getId()).size()).isEqualTo(4);
+
+        // Check historic identity-links when task is still active
+        List<HistoricIdentityLink> historicIdentityLinks = historyService.getHistoricIdentityLinksForTask(task.getId());
+        assertThat(historicIdentityLinks).hasSize(4);
+
+        // Validate all links
+        boolean foundCandidateUser = false, foundCandidateGroup = false, foundAssignee = false, foundCustom = false;
+        for (HistoricIdentityLink link : historicIdentityLinks) {
+            assertThat(link.getTaskId()).isEqualTo(task.getId());
+            if (link.getGroupId() != null) {
+                assertThat(link.getGroupId()).isEqualTo("sales");
+                foundCandidateGroup = true;
+            } else {
+                if (link.getType().equals("candidate")) {
+                    assertThat(link.getUserId()).isEqualTo("fozzie");
+                    foundCandidateUser = true;
+                } else if (link.getType().equals("assignee")) {
+                    assertThat(link.getUserId()).isEqualTo("kermit");
+                    foundAssignee = true;
+                } else if (link.getType().equals("actor")) {
+                    assertThat(link.getUserId()).isEqualTo("gonzo");
+                    assertThat(link.getDetails()).isEqualTo("details".getBytes());
+                    foundCustom = true;
+                }
+            }
+        }
+
+        assertThat(foundAssignee).isTrue();
+        assertThat(foundCandidateGroup).isTrue();
+        assertThat(foundCandidateUser).isTrue();
+        assertThat(foundCustom).isTrue();
```

Comment:
All this block can be simplified to
```suggestion
        assertThat(historicIdentityLinks)
            .extracting(
                HistoricIdentityLink::getType,
                HistoricIdentityLink::getGroupId,
                HistoricIdentityLink::getUserId,
                HistoricIdentityLink::getDetails)
            .containsExactlyInAnyOrder(
                tuple("candidate", "sales", null, null),
                tuple("candidate", null, "fozzie", null),
                tuple("assignee", null, "kermit", null),
                tuple("actor", null, "gonzo", "details".getBytes())
            );
```